src/courses/mappers/14.md

```diff
@@ -0,0 +1,1697 @@
+        'sample_jsons/SKELETON_mapper/sample_input_report/SKELETON.json',
+        {encoding: 'utf-8'}
+      ),
+      true
+    );
+
+    // fs.writeFileSync(
+    //   'sample_jsons/SKELETON_mapper/SKELETON-hdf-withraw.json',
+    //   JSON.stringify(mapper.toHdf(), null, 2)
+    // );
+
+    expect(omitVersions(mapper.toHdf())).toEqual(
+      omitVersions(
+        JSON.parse(
+          fs.readFileSync(
+            'sample_jsons/SKELETON_mapper/SKELETON-hdf-withraw.json',
+            {
+              encoding: 'utf-8'
+            }
+          )
+        )
+      )
+    );
+  });
+});
+```
+:::
+
+### Fingerprinting
+
+OHDF Converters has a fingerprinting service that detects a security tool data format and automatically applies the correct mapper to convert it to OHDF. To enable this feature, we need to explicitly declare keywords unique to the security tool data format as follows:
+
+1. Go to the file `report_intake.ts` under the `heimdall2/apps/frontend/src/store` directory.
+
+2. Import your mapper file. You should be able to add the name of your mapper class to a pre-existing import statement pointing at `hdf-converters` as follows:
+```typescript
+import {
+  ASFFResults as ASFFResultsMapper,
+  BurpSuiteMapper,
+  ...
+  {YOUR-MAPPER-CLASS-HERE}
+} from '@mitre/hdf-converters';
+```
+
+3. Instantiate your mapper class in the `convertToHdf` switch block. Add the following lines:
+```typescript
+case '{YOUR-EXPORT-SERVICE-NAME-HERE}':
+  return new {YOUR-MAPPER-CLASS-HERE}(convertOptions.data).toHdf();
+```
+
+4. Navigate to the file `fingerprinting.ts` in the `src/utils` directory in `hdf-converters`. Add keywords that are unique to your sample export to the `fileTypeFingerprints` variable. It should be formatted as follows:
+```typescript
+export const fileTypeFingerprints = {
+  asff: ['Findings', 'AwsAccountId', 'ProductArn'],
+  ...
+  {YOUR-EXPORT-SERVICE-NAME-HERE}: [{UNIQUE KEYWORDS AS STRINGS}]
+};
+```
+
+## Mapper Implementation
+
+With the necessary files now set up, we can begin the actual creation of the OHDF mapper using the skeleton mapper base in the `{YOUR-EXPORT-NAME-HERE}-mapper.ts` file. The skeleton mapper and the `base-converter` class have been designed to provide the base functionality needed for *-to-HDF mapper generation. For most developers, mapper creation will be limited to assigning objects from the export structure to correlating attributes in the mapper according to the mappings they developed earlier.
+
+::: warning File Processing
+ Certain security services produce exports which are not immediately usable by the skeleton mapper. In such case, pre-processing on the export and or post-processing on the generated OHDF file is necessary in order to ensure compatibility.
+:::
+
+While developing the mapper, it's useful to occassionally check on the outputed OHDF file to check that the mapper is working as intended. You can do this by starting a local instance of Heimdall2 with the following command:
```

Comment:
Might be worthwhile saying that the start:dev command will dynamically rebuild the application upon changes to the heimdall frontend but you're going to have to restart the command entirely to handle any changes to hdf converters.

Reply:
Addressed